### PR TITLE
【#562】fix context Keys map concurrent issue

### PR DESCRIPTION
**Pull request by UnderTreeTech**
1、fix context Keys map concurrent issue
2、add RemoteIP func to get client ip
3、add context pool

### CODE REVIEW

File: pkg/net/http/blademaster/context.go [hunk 1]
```diff
@@ -40,6 +44,8 @@ type Context struct {
 
 	// Keys is a key/value pair exclusively for the context of each request.
 	Keys map[string]interface{}
+	// This mutex protect Keys map
+	keysMutex *sync.RWMutex
```
Comment (tonyboxes) [inline pkg/net/http/blademaster/context.go:48]: keysMutex sync.RWMutex: remove pointer

File: pkg/net/http/blademaster/context.go [hunk 2]
```diff
@@ -93,16 +114,84 @@ func (c *Context) IsAborted() bool {
 // Set is used to store a new key/value pair exclusively for this context.
 // It also lazy initializes  c.Keys if it was not used previously.
 func (c *Context) Set(key string, value interface{}) {
+	if c.keysMutex == nil {
+		c.keysMutex = &sync.RWMutex{}
+	}
```
Comment (tonyboxes) [inline pkg/net/http/blademaster/context.go:119]: delete it, this is not thread-safe.

File: pkg/net/http/blademaster/context.go [hunk 3]
```diff
@@ -93,16 +114,84 @@ func (c *Context) IsAborted() bool {
...
 	c.Keys[key] = value
+	c.keysMutex.Unlock()
 }
 
 // Get returns the value for the given key, ie: (value, true).
 // If the value does not exists it returns (nil, false)
 func (c *Context) Get(key string) (value interface{}, exists bool) {
+	if c.keysMutex == nil {
+		c.keysMutex = &sync.RWMutex{}
+	}
```
Comment (tonyboxes) [inline pkg/net/http/blademaster/context.go:134]: delete it, this is not thread-safe.

File: pkg/net/http/blademaster/context.go [hunk 4]
```diff
@@ -51,6 +57,21 @@ type Context struct {
...
 
+/************************************/
+/********** CONTEXT CREATION ********/
+/************************************/
+func (c *Context) reset() {
+	c.Context = nil
+	c.index = -1
+	c.handlers = nil
+	c.Keys = nil
+	c.keysMutex = new(sync.RWMutex)
```
Comment (tonyboxes) [inline pkg/net/http/blademaster/context.go:68]: this lock does not need to be initialized.

File: pkg/net/http/blademaster/context.go (hunk 1 above)
Comment (UnderTreeTech) [reply pkg/net/http/blademaster/context.go]: okay, mutext default zero-value now.

**Comment by UnderTreeTech** [comment]
@tonyboxes errgroup test fail, pls fix it.

**Pull request merged**